File: DSA/src/Questions/Trie/PowerSet.java
package Questions.Trie;

import java.util.ArrayList;
import java.util.Collections;

public class PowerSet {
    static ArrayList<String> AllPossibleStrings(String s) {
        int n = s.length();
        ArrayList<String> ans = new ArrayList<>();
        for (int num = 0; num < (1 << n); num++) {
            String sub = "";
            for (int i = 0; i < n; i++) {
                // check if the ith bit is set or not
                if ((num & (1 << i)) != 0) {
                    sub += s.charAt(i);
                }
            }
            if (sub.length() > 0) {
                ans.add(sub);
            }
        }
        Collections.sort(ans);
        return ans;
    }
    // Time Complexity: O(2^n * n)

    // Reason: O(2^n) for the outer for loop and O(n) for the inner for loop.

    // Space Complexity: O(1)

    static void solveBackTracking(int i, String s, String f) {
        if (i == s.length()) {
            System.out.print(f + " ");
            return;
        }
        // picking
        // f = f + s.charAt(i);
        solveBackTracking(i + 1, s, f + s.charAt(i));
        // poping out while backtracking
        // f.pop_back();
        solveBackTracking(i + 1, s, f);
    }

    public static void main(String args[]) {

        String s = "abc";
        ArrayList<String> ans = AllPossibleStrings(s);
        // print all the subsequence.
        System.out.println("All possible subsequences are ");
        solveBackTracking(0, s, "");
        System.out.println();
        for (String it : ans) {
            System.out.print(it + " ");
        }

    }
}

File: DSA/Python/Questions/Trie/PowerSet.py
"""
# Question: Power Set
# Generate all possible subsets of a string

# Time Complexity: O(2^n * n)
# Space Complexity: O(1)

# Algorithm:
# 1. Use bit manipulation for subset generation
# 2. Generate all possible combinations
# 3. Track and sort subsets
# 4. Implement backtracking solution

# Key Components:
# - all_possible_strings(): Bit manipulation approach
# - solve_backtracking(): Recursive backtracking approach
"""


def all_possible_strings(s: str) -> list[str]:
    n = len(s)
    ans = []

    # Generate all possible combinations using bit manipulation
    for num in range(1, 1 << n):
        sub = ""
        for i in range(n):
            if num & (1 << i):
                sub += s[i]
        ans.append(sub)

    ans.sort()
    return ans


def solve_backtracking(i: int, s: str, f: str) -> None:
    if i == len(s):
        print(f, end=" ")
        return

    # Include current character
    solve_backtracking(i + 1, s, f + s[i])
    # Exclude current character
    solve_backtracking(i + 1, s, f)


def main():
    s = "abc"
    ans = all_possible_strings(s)

    print("All possible subsequences are:")
    solve_backtracking(0, s, "")
    print()

    print("All possible subsets are:")
    for subset in ans:
        print(subset, end=" ")


if __name__ == "__main__":
    main()
